### ISSUE TITLE
ROCm 3.9.1 fails to compile benchmark_warp_scan

### ISSUE THREAD

**Issue by MathiasMagnus** [ROCmSoftwarePlatform/rocPRIM#199]
**Describe the bug**
Compiler crashes when compiling `benchmark_warp_scan` with specific parameters.

**To Reproduce**
1. Install rocm-dkms version 3.9.1
2. Checkout rocPRIM `develop` at this [commit](https://github.com/ROCmSoftwarePlatform/rocPRIM/tree/fa69a09904e685b1aa236d57c975074e9749108d)
3. Set `HCC_AMDGPU_TARGET` env var to match CMake `AMDGPU_TARGETS`, as per [this](https://github.com/ROCm-Developer-Tools/HIP/issues/2181) issue. (otherwise you get unwanted targets in the binary too)
4. Configure
5. Build `benchmark_warp_scan`
3. See crash or open logs named `Build.log`

PowerShell script (port at your convenience) which enumerates all possible combinations:
```powershell
$configs = @(
     @('gfx803','gfx900','gfx906','gfx908'),
     @('gfx803','gfx900','gfx906'),
     @('gfx803','gfx900','gfx908'),
     @('gfx803','gfx906','gfx908'),
     @('gfx900','gfx906','gfx908'),
     @('gfx803','gfx900'),
     @('gfx803','gfx908'),
     @('gfx906','gfx908'),
     @('gfx900','gfx908'),
     @('gfx803','gfx906'),
     @('gfx803'),
     @('gfx900'),
     @('gfx906'),
     @('gfx908')
 );
$source_dir = "where you cloned your repo";
$build_root = "an existing empty folder";
foreach ($config in $configs) {
  $build_dir = ("$build_root/" + ($config -join "_"));
  New-Item -Type Directory $build_dir | Out-Null;
  $env:HCC_AMDGPU_TARGET = ($config -join ","); # hipcc expects comma delimited list
  cmake -DCMAKE_CXX_COMPILER=/opt/rocm/bin/hipcc -DCMAKE_BUILD_TYPE=Release -DBUILD_TEST=OFF -DBUILD_BENCHMARK=ON ("-DAMDGPU_TARGETS=" + ($config -join ";")) -S $source_dir -B $build_dir 2>&1 | Out-File $build_dir/Configure.log; # CMake expects semi-colon delimited list, need to guard from shell with quotes
  cmake --build $build_dir --target benchmark_warp_scan -- VERBOSE=1 2>&1 | Out-File $build_dir/Build.log;
}
```
These are the configurations that compile successfully:
```
# List folders that recursively somwhere contain the executable
(Get-ChildItem $build_root -Recurse -filter benchmark_warp_scan).Directory.Parent.BaseName
gfx900
gfx803
gfx908
gfx803_gfx908
gfx803_gfx900_gfx908
gfx803_gfx900_gfx906
gfx803_gfx906_gfx908
```
These are the configurations that _didn't_ compile successfully:
```
# Do the same, but this time, when do an extra `ls` of $build_root and exclude folders with the executables in them.
(Get-ChildItem $build_root -Exclude (gci $build_root -Recurse -filter benchmark_warp_scan).Directory.Parent.BaseName).BaseName
gfx803_gfx900
gfx803_gfx900_gfx906_gfx908
gfx803_gfx906
gfx900_gfx906_gfx908
gfx900_gfx908
gfx906
gfx906_gfx908
```

Note that moving to the [latest](https://github.com/ROCmSoftwarePlatform/rocPRIM/commit/4c65e18d3a55af8d2f78961803e8f20d4ce5140a) `develop` commit which [removed](https://github.com/ROCmSoftwarePlatform/rocPRIM/commit/36ce70ed43c98c03591a02a77c028a7ff750a2ab) `XNACK_FLAGS` in favor of more complex target names, the very same analysis can be done with a slightly altered script to omit funky path names:

```powerhshell
$configs = @(
@('gfx900:xnack-','gfx906:xnack-','gfx908:xnack-'),
@('gfx900:xnack-','gfx906:xnack-'),
@('gfx900:xnack-','gfx908:xnack-'),
@('gfx906:xnack-','gfx908:xnack-'),
@('gfx900:xnack-'),
@('gfx906:xnack-'),
@('gfx908:xnack-')
)
foreach ($config in $configs) {
$build_dir = ("$build_root/" + ($config -join "_").Replace(":xnack-",""));
...
}
```
Succesful builds:
```
gfx900                                                                       
gfx906                                            
gfx908                                                                               
gfx900_gfx908                                                                                                                                                                                                     gfx900_gfx906
```
Failing builds:
```
gfx900_gfx906_gfx908                                                         
gfx906_gfx908
```

**Expected behavior**
Compiler emitting a descriptive error if it was asked to do something impossible, or compile correctly otherwise.

**Log-files**
[This](https://github.com/ROCmSoftwarePlatform/rocPRIM/files/5536826/Build.txt) is the output of one such failure.
The funky arch names produce the same crash with a few extra `hipcc` and preprocessor warnings:
```
Warning: The specified HIP target: gfx906:xnack- is unknown. Correct compilation is not guaranteed.
Warning: The specified HIP target: gfx908:xnack- is unknown. Correct compilation is not guaranteed.
In file included from <built-in>:752:
<command line>:3:26: warning: ISO C99 requires whitespace after the macro name [-Wc99-extensions]
#define __HIP_ARCH_GFX906:XNACK-__ 1
                         ^
<command line>:4:26: warning: ISO C99 requires whitespace after the macro name [-Wc99-extensions]
#define __HIP_ARCH_GFX908:XNACK-__ 1
                         ^
```

**Environment**
[environment.txt](https://github.com/ROCmSoftwarePlatform/rocPRIM/files/5536838/environment.txt)

**Additional context**

**Comment by saadrahim** [comment]
Please use the tagged version from 3.9 release. Due to factors beyond my control, develop will not build on publicly available rocm version at the moment.

**Comment by xuhuisheng** [comment]
I think this issue could be solved after ROCm-4.0 released.

**Comment by saadrahim** [comment]
This has been fixed in the develop branch
